README.md
# MMM-OnThisDayWikiApi

[![License](https://img.shields.io/badge/license-MIT-blue.svg)](http://choosealicense.com/licenses/mit)

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/). It displays historical events from Wikipedia based on the current date using the free Wikimedia API. The event items will scroll through with the time-based (`updateInterval`).

## Screenshots

### Default Configuration

![Screenshot](screenshot/single_line.png)

### Configured to show as list

![Screenshot](screenshot/show_as_list.png)

## Install the module

Go to modules folder

```sh
cd modules
```

Clone this module from Github

```sh
git clone https://github.com/slugmuffin/MMM-OnThisDayWikiApi
```

After adding this module to your config (see below) restart your MagicMirror².

## Update the module

Go to modules folder

```sh
cd modules/MMM-OnThisDayWikiApi
```

Pull changes from Github

```sh
git pull
```

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:

```js
let config = {
    modules: [
        {
            module: "MMM-OnThisDayWikiApi",
            position: "bottom_bar", // All available positions
            config: {
                // See below for configurable options, this is optional
            },
        },
    ]
};
```

## Configuration options

All configuration options are _optional_ so the module works out of the box.

| Option | Description |
| ---------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `language` | Language code. <br><br> **Possible values:** <br> **NOTE:** *See below.* [Language](#language) <br> **Default value:** `en` |
| `type` | **Possible values:**<br>`all`: Returns all types<br>`selected`: Curated set of events that occurred on the given date<br>`births`: Notable people born on the given date<br>`deaths`: Notable people who died on the given date<br>`holidays`: Fixed holidays celebrated on the given date<br>`events`: Events that occurred on the given date that are not included in another type <br> **Default value:** `selected`|
| `showAsList` | Display the event items as a list. <br><br> **Possible values:** `true` or `false` <br> **Default value:** `false` |
| `maxDisplayList` | Total amount of items to display at once while cycling through (when shown as a list). <br><br> **Possible values:**`1` - `...` <br> **Default value:** `4` |
| `title` | The title to display (when shown as a list), enabled/disabled with `showTitle`. <br><br> **Possible values:** `string` <br> **Default value:** `"On This Day:"` |
| `showTitle` | Display the configured title (when shown as a list). <br><br> **Possible values:** `true` or `false` <br> **Default value:** `true` |
| `maxWidth` | Max width of the displayed events content (when shown as a list).<br>**Possible values:**`0` - `5000` <br> **Default value:**  `400`px |
| `wrapTitle` | Wrap the title of the item to multiple lines. <br><br> **Possible values:** `true` or `false` <br> **Default value:** `true` |
| `hideLoading` | Hide module instead of showing LOADING status. <br><br> **Possible values:** `true` or `false` <br> **Default value:** `false` |
| `reloadInterval` | How often does the content needs to be fetched? (Milliseconds) <br><br> **Possible values:** `1000` - `86400000` <br> **Default value:** `3600000` (1 hour) |
| `updateInterval` | How often do you want to display a new event? (Milliseconds) <br><br> **Possible values:**`1000` - `60000` <br> **Default value:** `10000` (10 seconds) |
| `animationSpeed` | Speed of the update animation. (Milliseconds) <br><br> **Possible values:**`0` - `5000` <br> **Default value:** `1000` (1 seconds) |
| `sortItems` | Whether or not to sort the items by year recent->past. <br><br> **Possible values:**`true` - `false` <br> **Default value:** `true` |
| `maxItems` | Limits the total amount of items stored (after sort). (0 for unlimited) <br><br> **Possible values:**`0` - `...` <br> **Default value:** `0` |
| `prohibitedWords` | Remove item if one of these words is found anywhere in the title (case insensitive and greedy matching) <br><br> **Possible values:** `['word']` or `['word1','word2',...]` |

## Language

Supports the same languages as the Wikimedia API.
To translate/customize the `title`, use the `config`.

**Defaults to english.**

Currently supported languages:
| Language  | Language   | code |
| --------- | ---------- | ---- |
| English   | English    | en   |
| Deutsch   | German     | de   |
| Français  | French     | fr   |
| Svenska   | Swedish    | sv   |
| Português | Portuguese | pt   |
| Русский   | Russian    | ru   |
| Español   | Spanish    | es   |
| العربية     | Arabic     | ar   |
| Bosanski  | Bosnian    | bs   |

## Problems

If you have any problems or questions, feel free to open an issue. There are many possible improvements for this module so please let me know if you miss something.

## License: MIT

See [LICENSE](LICENSE.txt)

## Thanks

Thank you to:

- [MagicMirror²](https://github.com/MichMich/MagicMirror/) (specifically `newsfeed`)
- [MMM-OnThisDay](https://github.com/nkl-kst/MMM-OnThisDay)
- [MMM-DailyBibleVerse](https://github.com/arthurgarzajr/MMM-DailyBibleVerse)
- [on-this-day](https://github.com/elliefairholm/on-this-day)


CHANGELOG.md
# MMM-OnThisDay changelog

All notable changes to this project will be documented in this file.

## [0.2.0] 2023-12-13

- Replace `request` by internal fetch - with that `npm install` isn't necessary anymore.
- Fix typos and format

## [0.1.0] 2023-12-12

- Initial version.


node_helper.js
/* MagicMirror²
 * Node Helper: MMM-OnThisDayWikiApi
 *
 * By Ryan Seeber
 * MIT Licensed.
 */

const NodeHelper = require("node_helper");

module.exports = NodeHelper.create({
  // Subclass start method.
  start() {
    console.log("Started node_helper.js for MMM-OnThisDayWikiApi.");
  },

  socketNotificationReceived(notification, payload) {
    console.log(
      `${this.name} node helper received a socket notification: ${notification} - Payload: ${payload}`
    );
    if (notification === "ONTHISDAY_FETCH") {
      this.wikiOnThisDayRequest(
        payload.config.apiBase,
        payload.config.language,
        payload.config.feature,
        payload.config.type,
        payload.month,
        payload.day
      );
    } else {
      console.log(
        `${this.name} node helper ERROR: received unknown notification.`
      );
    }
  },

  async wikiOnThisDayRequest(apiBase, language, feature, type, month, day) {
    const self = this;

    const url = `${apiBase}/${language}/${feature}/${type}/${month}/${day}`;
    console.log(url);

    try {
      const response = await fetch(url, { method: "GET" });
      if (response.ok) {
        const wikiData = await response.json();
        const items = wikiData[type];
        if (items.length <= 0) {
          console.log("OnThisDay-Fetcher: No items to broadcast yet.");
          return;
        }
        console.log(`OnThisDay-Fetcher: Broadcasting ${items.length} items.`);
        self.sendSocketNotification("ONTHISDAY_ITEMS", items);
      } else if (response.status === 400) {
        const errorMsg = "Invalid parameter";
        console.error(errorMsg);
        self.sendSocketNotification("ONTHISDAY_ERROR", {
          error_type: errorMsg
        });
      } else if (response.status === 404) {
        const errorMsg = "No data found for the requested date";
        console.error(errorMsg);
        self.sendSocketNotification("ONTHISDAY_ERROR", {
          error_type: errorMsg
        });
      } else if (response.status === 501) {
        const errorMsg = "Unsupported language";
        console.error(errorMsg);
        self.sendSocketNotification("ONTHISDAY_ERROR", {
          error_type: errorMsg
        });
      } else {
        const errorMsg = `Bad response from server: ${response.status}`;
        console.error(errorMsg);
        self.sendSocketNotification("ONTHISDAY_ERROR", {
          error_type: errorMsg
        });
      }
    } catch (error) {
      const errorMsg = "Unknown error";
      console.error(errorMsg, error);
      self.sendSocketNotification("ONTHISDAY_ERROR", { error_type: errorMsg });
    }
  }
});


MMM-OnThisDayWikiApi.js
/* MagicMirror²
 * Module: MMM-OnThisDayWikiApi
 *
 * By Ryan Seeber
 * MIT Licensed.
 */

Module.register("MMM-OnThisDayWikiApi", {
  // Default module config.
  result: [],
  defaults: {
    apiBase: "https://api.wikimedia.org/feed/v1/wikipedia",
    feature: "onthisday",
    language: "en",
    type: "selected",
    showAsList: false,
    maxDisplayList: 4,
    title: "On this day in",
    showTitle: true,
    maxWidth: 400,
    wrapTitle: true,
    hideLoading: true,
    reloadInterval: 60 * 60 * 1000, // ms (1 hour)
    updateInterval: 10 * 1000, // ms (10 seconds)
    animationSpeed: 1000,
    sortItems: true,
    maxItems: 0,
    prohibitedWords: []
  },

  start() {
    Log.info(`Starting module: ${this.name}`);
    const self = this;

    this.loaded = false;
    this.activeItem = 0;
    this.scrollPosition = 0;
    this.onThisDayItems = [];

    const timeStamp = new Date();
    const monthStr = (timeStamp.getMonth() + 1).toString().padStart(2, "0");
    const dayStr = timeStamp.getDate().toString().padStart(2, "0");

    // Set up a fetch for hourly
    setInterval(function () {
      Log.debug(`${this.name} Fetch new OnThisDay data`);

      self.sendSocketNotification("ONTHISDAY_FETCH", {
        config: self.config,
        month: monthStr,
        day: dayStr
      });
    }, this.config.reloadInterval); // perform every hour (3600000 milliseconds)

    // But get it populted right now
    self.sendSocketNotification("ONTHISDAY_FETCH", {
      config: self.config,
      month: monthStr,
      day: dayStr
    });
  },

  /**
   * Schedule visual update.
   */
  scheduleUpdateInterval() {
    this.updateDom(this.config.animationSpeed);

    // #2638 Clear timer if it already exists
    if (this.timer) clearInterval(this.timer);

    this.timer = setInterval(() => {
      let incrementValue = 1;
      if (this.config.showAsList) {
        incrementValue = this.config.maxDisplayList;
      }
      this.activeItem += incrementValue;
      Log.debug(`${this.name} this.activeItem = " ${this.activeItem}`);
      this.updateDom(this.config.animationSpeed);
    }, this.config.updateInterval);
  },

  getStyles() {
    return ["MMM-OnThisDayWikiApi.css"];
  },

  // Override fetching of template name
  getTemplate() {
    Log.debug(`${this.name} - getTemplate`);
    return "MMM-OnThisDayWikiApi.njk";
  },

  getHeader() {
    return null;
  },

  // Override template data and return whats used for the current template
  getTemplateData() {
    Log.debug(`${this.name} - getTemplateData`);
    if (this.error) {
      return {
        error: this.error
      };
    }

    if (this.onThisDayItems.length === 0) {
      return {
        empty: true
      };
    }
    if (this.activeItem >= this.onThisDayItems.length) {
      this.activeItem = 0;
      Log.debug(`${this.name} this.activeItem = " ${this.activeItem}`);
    }

    let sliceStart = this.activeItem;
    let sliceEnd = this.activeItem + this.config.maxDisplayList;

    // Always show the same same amount
    if (sliceEnd > this.onThisDayItems.length) {
      sliceEnd = this.onThisDayItems.length;
      sliceStart = this.onThisDayItems.length - this.config.maxDisplayList;
    }

    Log.debug(`${this.name} slice_start: ${sliceStart} slice_end: ${sliceEnd}`);

    const item = this.onThisDayItems[this.activeItem];
    const items = this.onThisDayItems.slice(sliceStart, sliceEnd);
    Log.debug(items);

    return {
      loaded: true,
      empty: false,
      config: this.config,
      activeIndex: this.activeItem,
      activeItem: item,
      items: items
    };
  },

  socketNotificationReceived(notification, payload) {
    if (notification === "MODULE_DOM_CREATED" && this.config.hideLoading) {
      this.hide();
    } else if (notification === "ONTHISDAY_ITEMS") {
      this.generateFeed(payload);

      if (!this.loaded) {
        Log.debug(`${this.name} - wasn't loaded yet.`);
        this.loaded = true;
        this.error = null;

        if (this.config.hideLoading) {
          this.show();
        }
        this.scheduleUpdateInterval();
        Log.debug(`${this.name} - rotation updates scheduled`);
      }

      this.loaded = true;
      this.error = null;
    } else if (notification === "ONTHISDAY_ERROR") {
      this.error = this.translate(payload.error_type);
      Log.error(`${this.name} - ${this.error}`);
      this.scheduleUpdateInterval();
    }
  },

  /**
   * Generate an ordered list of items for this configured module.
   * @param {object} onThisDayItems An object with feeds returned by the node helper.
   */
  generateFeed(onThisDayItems) {
    Log.debug(`${this.name} - generateFeed`);

    if (this.config.sortItems) {
      onThisDayItems.sort(function (a, b) {
        return b.year - a.year;
      });
    }

    if (this.config.maxItems > 0) {
      onThisDayItems = onThisDayItems.slice(0, this.config.maxItems);
    }

    if (this.config.prohibitedWords.length > 0) {
      onThisDayItems = onThisDayItems.filter(function (item) {
        for (const word of this.config.prohibitedWords) {
          if (item.title.toLowerCase().indexOf(word.toLowerCase()) > -1) {
            return false;
          }
        }
        return true;
      }, this);
    }

    this.onThisDayItems = onThisDayItems;
  }
});
